django/contrib/admindocs/utils.py

```diff
@@ -151,90 +154,188 @@ def replace_metacharacters(pattern):
     )
 
 
-def replace_named_groups(pattern):
-    r"""
-    Find named groups in `pattern` and replace them with the group name. E.g.,
-    1. ^(?P<a>\w+)/b/(\w+)$ ==> ^<a>/b/(\w+)$
-    2. ^(?P<a>\w+)/b/(?P<c>\w+)/$ ==> ^<a>/b/<c>/$
-    3. ^(?P<a>\w+)/b/(\w+) ==> ^<a>/b/(\w+)
-    4. ^(?P<a>\w+)/b/(?P<c>\w+) ==> ^<a>/b/<c>
+def _find_groups(pattern, group_matcher, named):
+    """
+    Protected helper function to return start and end indices of a group in the
+    give pattern and if `named` is `True`, also returns the `group name` inside
+    the pattern.
+
+    Note: This function does not do any sanity check on the arguments it takes
+    in because this is meant to be a part of implementation so please make sure
+    to pass in valid arguments to this function.
+
+    Args:
+        pattern:        `Regex` pattern for a `named`, `unnamed` or
+                        `non-capturing` group.
+        group_matcher:  Pre-compiled `regex` object to match either the `named`,
+                        `unnammed` or `non-capturing` group inside the `pattern`.
+        named:          If `True` then influence the behaviour of this function
+                        by directing it to follow a certain set of instructions
+                        to generate a `group_start_end_indices` list for `named`
+                        groups.
+
+    Returns:
+        List of `tuples` containing the `start` and `end` indices of all the
+        groups present inside of the given `pattern`.  The `tuple` will contain
+        a third extra element i.e., 'group name' if `named` is set to `True`;
+        this is a need inside of the function `replace_named_groups()` to
+        replace the entire named groups in `pattern` with the 'group name'.
     """
-    named_group_indices = [
-        (m.start(0), m.end(0), m[1])
-        for m in named_group_matcher.finditer(pattern)
-    ]
-    # Tuples of (named capture group pattern, group name).
-    group_pattern_and_name = []
-    # Loop over the groups and their start and end indices.
-    for start, end, group_name in named_group_indices:
-        # Handle nested parentheses, e.g. '^(?P<a>(x|y))/b'.
+    if named is True:
+        unnamed_group_indices = [
+            (m.start(0), m.end(0), m[1])
+            for m in group_matcher.finditer(pattern)
+        ]
+    else:
+        unnamed_group_indices = [(m.start(0), m.end()) for m in group_matcher.finditer(pattern)]
```

Comment:
We don't need to add `named` branches everywhere. I'd return `m` in the tuple and use it when needed. Moreover, we could use a generator instead of list.

IMO, `_find_group()` is unnecessarily complicated and unreadable.

I'd move `get_group_start_and_end_indices()` to a module hook `_get_group_start_end()`:
```python
def _get_group_start_end(start, end, pattern):
    unmatched_open_brackets, prev_char = 1, None
    for idx, val in enumerate(pattern[end:]):
        if val == '(' and prev_char != '\\':
            unmatched_open_brackets += 1
        elif val == ')' and prev_char != '\\':
            unmatched_open_brackets -= 1
        prev_char = val
        if unmatched_open_brackets == 0:
            return start, (end + idx + 1)
```
and simplify `_find_groups()` to a few lines generator:
```python
def _find_groups(pattern, group_matcher):
    prev_end = None
    for m in group_matcher.finditer(pattern):
        if indices := _get_group_start_end(m.start(0), m.end(0), pattern):
            start, end = indices
            if prev_end and start > prev_end or not prev_end:
                yield start, end, m
            prev_end = end
```
now you can use `m` when needed, e.g.:
```python
def replace_named_groups(pattern):
    group_pattern_and_name = [
        (pattern[start:end], match[1])
        for start, end, match in _find_groups(pattern, named_group_matcher)
    ]
    for group_pattern, group_name in group_pattern_and_name:
        pattern = pattern.replace(group_pattern, group_name)
    return pattern
...
def replace_unnamed_groups(pattern):
    group_start_end_indices = _find_groups(pattern, unnamed_group_matcher)
    final_pattern, prev_end = '', None
    for start, end, _ in group_start_end_indices:
        if prev_end:
            final_pattern += pattern[prev_end:start]
        final_pattern += pattern[:start] + '<var>'
        prev_end = end
    return final_pattern + pattern[prev_end:]
...

```